### PR TITLE
v4 Upgrade Guide

### PR DESCRIPTION
**Pull request by nwithan8**
# Description

- Add upgrade guide for v3->v4

# Testing

- N/A

# Pull Request Type

Please select the option(s) that are relevant to this PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement (fixing a typo, updating readme, renaming a variable name, etc)

### CODE REVIEW

**Comment by nwithan8** [comment]
Linting fails, requires this PR: https://github.com/EasyPost/easypost-csharp/pull/339

File: UPGRADE_GUIDE.md [hunk 1]
```diff
@@ -2,7 +2,203 @@
...
+
+- [Updating Dependencies](#40-updating-dependencies)
+- [Client Instance](#40-client-instance)
+- [Language Conventions](#40-language-conventions)
+- [Nullable Properties](#40-nullable-properties)
+
+### 4.0 Medium Impact Changes
+
+- [Error Types](#40-error-types)
+- [Beta Client](#40-beta-client)
```
Comment (Justintime50) [inline UPGRADE_GUIDE.md:20]: I'd say the beta client is low since so few customers are beta product users.

File: UPGRADE_GUIDE.md [hunk 2]
```diff
@@ -2,7 +2,203 @@
...
+### 4.0 High Impact Changes
+
+- [Updating Dependencies](#40-updating-dependencies)
+- [Client Instance](#40-client-instance)
+- [Language Conventions](#40-language-conventions)
+- [Nullable Properties](#40-nullable-properties)
+
+### 4.0 Medium Impact Changes
+
+- [Error Types](#40-error-types)
```
Comment (Justintime50) [inline UPGRADE_GUIDE.md:19]: I'd argue error handling is high impact since everyone deals with them.

File: UPGRADE_GUIDE.md [hunk 3]
```diff
@@ -2,7 +2,203 @@
...
+- [Services and Resources](#40-services-and-resources)
+
+## 4.0 Updating Dependencies
+
+*Likelihood of Impact: **High***
+
+**Dependencies**
+
+- RestSharp was upgraded from v107 to v108
+- All dependencies had minor version bumps
```
Comment (Justintime50) [inline UPGRADE_GUIDE.md:33]: Thought: I'd maybe also callout that all deps are now locked to a major version.

File: UPGRADE_GUIDE.md [hunk 4]
```diff
@@ -2,7 +2,203 @@
...
+```
+
+All functions are now accessed via the `Client` instance. For example, to create a shipment:
+
+```csharp
+// Old
+EasyPost.Shipment shipment = await EasyPost.Shipment.Create(parameters);
+
+// New
+EasyPost.Shipment shipment = awiat myClient.Shipment.Create(parameters);
```
Comment (Justintime50) [inline UPGRADE_GUIDE.md:61]: Issue: 
```suggestion
EasyPost.Shipment shipment = await myClient.Shipment.Create(parameters);
```

File: UPGRADE_GUIDE.md [hunk 5]
```diff
@@ -2,7 +2,203 @@
...
+
+The majority of properties for an object will likely be populated as a result of an API call, but if not, end-users will
+know that the value of a non-populated property is guaranteed to be `null`.
+
+End users may need to add additional null checks or nullability markers in their code, although this change will likely
+play well with any existing null checks.
+
+```csharp
+// Old
+string id = myAddress.id; // "id" is non-nullable, but not guaranteed to be populated
```
Comment (Justintime50) [inline UPGRADE_GUIDE.md:100]: Issue: ID is always guaranteed to be populated heh. Let's choose something that is not guaranteed to be nullable so the example is more appropriate. This as written is confusing.

File: UPGRADE_GUIDE.md [hunk 6]
```diff
@@ -2,7 +2,203 @@
...
+bool correctId = id?.startsWith("adr_") ?? false; // no compilation warning
+```
+
+## 4.0 Error Types
+
+*Likelihood of Impact: **Medium***
+
+Error handling has been overhauled and a number of specific exception types have been introduced.
+
+All exceptions inherit from `EasyPost.Exceptions.EasyPostError`. Users can catch this exception type to handle all
```
Comment (Justintime50) [inline UPGRADE_GUIDE.md:114]: Thought: We need to call out that we are also extending the builtin C# exception right?

File: UPGRADE_GUIDE.md [hunk 7]
```diff
@@ -2,7 +2,203 @@
...
+- `string? Code` - the error code returned by the API (e.g. "PARAMETER.INVALID")
+- `List<EasyPost.Error>? Errors` - a list of errors returned by the API (e.g. "Invalid parameter: to_address")
+
+Any `EasyPost.Exceptions.ApiError` exception can be pretty-printed using the `PrettyPrint` string property.
+
+Users can better anticipate exception information through utilities in the `EasyPost.Exceptions.Constants` namespace.
+
+- Check what exception type will be thrown for a given HTTP status code by
+  using `EasyPost.Exceptions.Constants.GetEasyPostExceptionType(HttpStatusCode statusCode)`
+- Check error message strings/templates in `EasyPost.Exceptions.Constants.ErrorMessages`
```
Comment (Justintime50) [inline UPGRADE_GUIDE.md:169]: Would like some examples of usage for this please

File: UPGRADE_GUIDE.md [hunk 8]
```diff
@@ -2,7 +2,203 @@
...
+
+## 4.0 Beta Client
```
Comment (Justintime50) [inline UPGRADE_GUIDE.md:171]: Thought: "Beta Client" is misleading. I'd just call it "namespace" or "property" instead

File: UPGRADE_GUIDE.md [hunk 9]
```diff
@@ -2,7 +2,203 @@
...
+
+```csharp
+// Old
+var betaObject = await EasyPost.Beta.SomeService.SomeFunction(parameters);
+
+// New
+var betaObject = await myClient.Beta.SomeService.SomeFunction(parameters);
+```
+
+Any beta function will use the same client (and associated API key) from which the beta function was accessed (
```
Comment (Justintime50) [inline UPGRADE_GUIDE.md:186]: Thought: I'd add info here detailing that this client can/should only be used for beta functions and that a separate client should be made in the general namespace for everything else.

**Review by Justintime50 — DISMISSED** [review]

File: UPGRADE_GUIDE.md (hunk 1 above)
Comment (nwithan8) [reply UPGRADE_GUIDE.md:20]: Resolved

File: UPGRADE_GUIDE.md (hunk 2 above)
Comment (nwithan8) [reply UPGRADE_GUIDE.md:19]: Resolved

File: UPGRADE_GUIDE.md (hunk 3 above)
Comment (nwithan8) [reply UPGRADE_GUIDE.md:33]: Resolved

File: UPGRADE_GUIDE.md (hunk 4 above)
Comment (nwithan8) [reply UPGRADE_GUIDE.md:61]: Resolved

File: UPGRADE_GUIDE.md (hunk 5 above)
Comment (nwithan8) [reply UPGRADE_GUIDE.md:100]: Resolved

File: UPGRADE_GUIDE.md (hunk 6 above)
Comment (nwithan8) [reply UPGRADE_GUIDE.md:114]: I would imagine for a .NET dev that would be assumed, but I will call it out. Resolved.

File: UPGRADE_GUIDE.md (hunk 8 above)
Comment (nwithan8) [reply UPGRADE_GUIDE.md:171]: Resolved

File: UPGRADE_GUIDE.md (hunk 9 above)
Comment (nwithan8) [reply UPGRADE_GUIDE.md:186]: Resolved, although this isn't really an issue. The beta client only has access to beta features (you can't call a non-beta feature off of a beta client). If you somehow did pass a beta client into a non-beta API call, it wouldn't matter (the beta client isn't actually what dictates the beta endpoint, that's the functions themselves. The beta client really is just a clone of the normal client.)

File: UPGRADE_GUIDE.md (hunk 7 above)
Comment (nwithan8) [reply UPGRADE_GUIDE.md:169]: What did you have in mind?

File: UPGRADE_GUIDE.md (hunk 6 above)
Comment (Justintime50) [reply UPGRADE_GUIDE.md:114]: Well... to this point many of our libs don't extend the builtin error/exception 😂 So just making sure people know if they need that info (think Node which is sad)

File: UPGRADE_GUIDE.md (hunk 9 above)
Comment (Justintime50) [reply UPGRADE_GUIDE.md:186]: Ah yes, thanks

File: UPGRADE_GUIDE.md (hunk 7 above)
Comment (Justintime50) [reply UPGRADE_GUIDE.md:169]: Mmm re-reading this it's fine.

**Comment by Justintime50** [comment]
Linter is sad, needs to be rebased.

**Review by Justintime50 — APPROVED** [review]

**Review by jchen293 — APPROVED** [review]

**Review by Justintime50 — APPROVED** [review]
Re-LGTM

**Pull request merged**